### ISSUE TITLE
Feature request: localize sudo

### ISSUE THREAD

**Issue by rouson** [ParaToolsInc/taucmdr#223]
Kudos on the simple installation process.  I love it!

Observed behavior
-------------------
The installation fails when executing `make install INSTALLDIR=...` with a directory that requires `sudo` privileges.  Prepending `sudo` to `make` appears to be working, but this means the whole installation process has `sudo` privileges rather than just the steps that need it.

Desired behavior
------------------
It would be nice if `sudo` could be used only at the points when it is necessary.  In OpenCoarrays, we do this in a bash script by defining a `SUDO` variable that contains an empty string if the installation directory does not require `sudo` and contains the full `sudo` command (which sometimes is longer than just `sudo`) otherwise.  See [set_SUDO_if_needed_to_write_to_directory.sh].  Then the user gets a password prompt, if necessary, at the point(s) of invoking `sudo`.  The trade-off is that this precludes guaranteeing a completely non-interactive build process, but the only required interaction comes near the end of the installation process when the installation directory is being created and files are being installed.

[set_SUDO_if_needed_to_write_to_directory.sh]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/prerequisites/build-functions/set_SUDO_if_needed_to_write_to_directory.sh

**Comment by zbeekman** [comment]
Hi Damian,

Can you explain your objection to using `sudo make install INSTALLDIR=...` in more detail? I'm not sure I understand why `sudo make install ...` is undesirable for TAU Commander.

TAU Commander exists due to the very complex process of downloading configuring and installing TAU and all the other dependencies which are also a function of the system (AKA target),  application (does it use MPI? OpenMP? etc.) and the measurement(s) you want to perform (Sampling vs profiling vs tracing and the last two with or without PAPI metrics, IO profiling etc.). TAU Commander tries to minimize the amount of time spent downloading, configuring and compiling software between different users and projects/applications. Correctly determining the permissions for the tau/taucmdr installation is a non-trivial process and `sudo make install ...` achieves this, as far as I know. If you take a look at [the Makefile](https://github.com/ParaToolsInc/taucmdr/blob/master/Makefile) you'll see that all of the heavy lifting and actual installation logic is handled by Python after installing the miniconda python distribution.

Also, just FYI, only Linux is fully supported. macOS can be used for visualization with ParaProf, but collecting performance data on macOS is not currently supported.

**Comment by zbeekman** [comment]
OK, I just re-read your comment and it is making more sense to me now.

To clarify: Are you asking for an interactive prompt during the install process to give sudo privileges to the installation in the case that the user forgets to specify `sudo` when running `make install ...`?

If so this is a fairly straightforward fix/request, I believe.

**Comment by rouson** [comment]
In response to your clarification, yes, that's exactly it.  I wrote aforementioned script in OpenCoarrays at the request of @tclune. I think the motivation is twofold:

1. If the entire process runs with `sudo` privileges, then removing many of the created files (which could include more than just the installed files) also requires `sudo`. The more people have to use `sudo`, the greater is the risk of mistakenly mis-using it.
2. If the entire process runs with sudo privileges, then the potential for a costly mistake is higher.

Those are just my guesses as to why Tom made the request.  There could be other reasons that I haven't considered.

**Comment by tclune** [comment]
Actually, the reason is simpler.    I no longer have sudo privileges on my machine.
NASA is gradually tightening the knobs on that ability.

In theory I can still get such privileges for 30 days at a time.  And even longer if I then hide from the sysadmin :-)
Not sure why temporary privileges are thought to be safer than permanent.   But I’ve essentially been forced to
find ways to live without sudo.    I do get the sysadmins to install commercial software, but something as quickly
evolving as OpenCoarrays would be tedious to update in that mode.


On Oct 27, 2017, at 9:02 AM, Damian Rouson <notifications@github.com<mailto:notifications@github.com>> wrote:


In response to your clarification, yes, that's exactly it. I wrote aforementioned script in OpenCoarrays at the request of @tclune<https://github.com/tclune>. I think the motivation is twofold:

  1.  If the entire process runs with sudo privileges, then removing many of the created files (which could include more than just the installed files) also requires sudo. The more people have to use sudo, the greater is the risk of mistakenly mis-using it.
  2.  If the entire process runs with sudo privileges, then the potential for a costly mistake is higher.

Those are just my guesses as to why Tom made the request. There could be other reasons that I haven't considered.

**Comment by jlinford** [comment]
A reminder to use `sudo` is easy, but I think it's not actually what you want.

TL;DR: The environment during TAU installation must exactly match the environment under which TAU will be used.

During installation, the TAU Performance System (not TAU Commander, but the core, underlying TAU from University of Oregon) does many things which are now considered bad practice.  It modifies files in place, copies files to the installation directory *then* compiles them, and retains state between `make install` invocations.  If you're installing TAU in a location that is only writable by superuser then *all* installation actions must be performed by the superuser.  This is unavoidable.

This problem is compounded by differences between the superuser's environment and the typical user's.  For example, if root has `CC=/usr/bin/gcc` but most users have `CC=/opt/gcc/bin/gcc` then installing TAU as root will configure TAU with `/usr/bin/gcc` instead of the desired `/opt/gcc/bin/gcc`.  This is one of the reasons why the global installations of the TAU Performance System typically only work for one or two lucky users.

TAU Commander solves these problems by installing TAU and its dependencies in user space when it detects that the system-level TAU installation differs from the user's environment.  **You do not need superuser privileges to use TAU Commander,** just enough space in your home directory (or somewhere) to install TAU and its dependencies.

So you have two options:
  1) [Recommended] Grant a non-root user write privilege on INSTALLDIR, install taucmdr as non-root user using the typical user's environment, then revoke write privilege after installation completes.  This is what ParaTools does at client sites.

  2) Install everything as root, i.e. `sudo make install`.  This is not recommended because even if installation succeeds you'll have TAU configured for the root user's environment, not the typical user's environment.

**Comment by rouson** [comment]
Thanks for the quick replies!  I'll close this issue.  Feel free to reopen it if there's anything else to address.
